framework/src/media/MediaRecorderImpl.cpp

```diff
@@ -136,11 +136,13 @@ void MediaRecorderImpl::prepareRecorder(recorder_result_t& ret)
 	if (mCurState != RECORDER_STATE_IDLE || mOutputDataSource == nullptr) {
 		meddbg("prepare Failed mCurState: %d mOutputDataSource : %s\n", (recorder_state_t)mCurState, \
 			(mOutputDataSource == nullptr ? "nullptr" : "not null"));
+		ret = (mOutputDataSource == nullptr) ? RECORDER_INVALID_DATASOURCE : RECORDER_INVALID_STATE;
```

Comment:
I forgot this for a long time. we should call setDatasource as a mandatory, then make one more state for datasource.
No this time, in past future

Reply:
yes, it is mandatory.
but, user can set the nullptr possible. 
Do we need more than RECORDER_INVALID_DATASOURCE state?

Reply:
You can refer to below, DataSourceConfigured state.
https://developer.android.com/reference/android/media/MediaRecorder